# Insertion-paired Read Alignment Workflow

### Inputs
# Alter config.yaml to provide the following:
 * **Sample name**: names the output directory

 * **Reads 1, Reads 2**: forward and reverse reads in fastq or fastq.gz format

 * **Insertion**: Sequences to locate within the assembly with paired read evidence.  Fasta format.

 * **Assembly**: Sequence within which to locate the unseen duplicated sequences.  Fasta format.
